adapters/insticator/insticator.go

```diff
@@ -0,0 +1,316 @@
+	}
+	for _, seatBid := range response.SeatBid {
+		for i := range seatBid.Bid {
+			bid := &seatBid.Bid[i]
+			bidType := getMediaTypeForBid(bid)
+			b := &adapters.TypedBid{
+				Bid:     &seatBid.Bid[i],
+				BidType: bidType,
+			}
+			bidResponse.Bids = append(bidResponse.Bids, b)
+		}
+	}
+	return bidResponse, nil
+}
+
+func makeImps(imp openrtb2.Imp) (openrtb2.Imp, error) {
+
+	var bidderExt adapters.ExtImpBidder
+	if err := json.Unmarshal(imp.Ext, &bidderExt); err != nil {
+		return openrtb2.Imp{}, &errortypes.BadInput{
+			Message: err.Error(),
+		}
+	}
+
+	var insticatorExt openrtb_ext.ExtImpInsticator
+	if err := json.Unmarshal(bidderExt.Bidder, &insticatorExt); err != nil {
+		return openrtb2.Imp{}, &errortypes.BadInput{
+			Message: err.Error(),
+		}
+	}
+
+	var impExt ext
+	impExt.Insticator.AdUnitId = insticatorExt.AdUnitId
+	impExt.Insticator.PublisherId = insticatorExt.PublisherId
+
+	impExtJSON, err := json.Marshal(impExt)
+	if err != nil {
+		return openrtb2.Imp{}, &errortypes.BadInput{
+			Message: err.Error(),
+		}
+	}
+
+	imp.Ext = impExtJSON
+	// Validate Video if it exists
+	if imp.Video != nil {
+		videoCopy, err := validateVideoParams(imp.Video)
+
+		imp.Video = videoCopy
+
+		if err != nil {
+			return openrtb2.Imp{}, &errortypes.BadInput{
+				Message: err.Error(),
+			}
+		}
+	}
+
+	return imp, nil
+}
+
+func makeReqExt(request *openrtb2.BidRequest) ([]byte, error) {
+	var reqExt reqExt
+
+	if len(request.Ext) > 0 {
+		if err := json.Unmarshal(request.Ext, &reqExt); err != nil {
+			return nil, err
+		}
+	}
+
+	if reqExt.Insticator == nil {
+		reqExt.Insticator = &reqInsticatorExt{}
+	}
+
+	if reqExt.Insticator.Caller == nil {
+		reqExt.Insticator.Caller = make([]insticatorCaller, 0)
+	}
+
+	reqExt.Insticator.Caller = append(reqExt.Insticator.Caller, caller)
+
+	return json.Marshal(reqExt)
+}
+
+func resolveBidFloor(bidFloor float64, bidFloorCur string, reqInfo *adapters.ExtraRequestInfo) (float64, error) {
+	if bidFloor > 0 && bidFloorCur != "" && strings.ToUpper(bidFloorCur) != "USD" {
+		floor, err := reqInfo.ConvertCurrency(bidFloor, bidFloorCur, "USD")
+		return roundTo4Decimals(floor), err
+	}
+
+	return bidFloor, nil
+}
+
+// roundTo4Decimals function
+func roundTo4Decimals(amount float64) float64 {
+	return math.Round(amount*10000) / 10000
+}
+
+func validateVideoParams(video *openrtb2.Video) (*openrtb2.Video, error) {
+	videoCopy := *video
+	if (videoCopy.W == nil || *videoCopy.W == 0) ||
+		(videoCopy.H == nil || *videoCopy.H == 0) ||
+		videoCopy.MIMEs == nil {
+
+		return nil, &errortypes.BadInput{
+			Message: "One or more invalid or missing video field(s) w, h, mimes",
+		}
+	}
+
+	return &videoCopy, nil
+}
+
+// populate publisherId to site/app object from imp extension
+func populatePublisherId(imp *openrtb2.Imp, request *openrtb2.BidRequest) {
+	var ext ext
+
+	if request.Site != nil && request.Site.Publisher == nil {
+		request.Site.Publisher = &openrtb2.Publisher{}
```

Comment:
You need to clone `request.Site` before changing it's fields, as that is shared memory. You are likely missing json test coverage for this to not be picked up in failing tests.

Same for line 309 in this PR for `request.App`.

Reply:
I see, @SyntaxNode . I can clone the `request.Site`/ `request.App` and then set the `publisher.id`. But are you suggesting to add json test cases too for the same?

I did add the test cases to validate `publisher.id`.

Reply:
@SyntaxNode updated the PR.